src/zeit/content/cp/automatic.py

```diff
@@ -67,116 +54,170 @@ def values(self):
-        """
-        result = None
-        pop = []
-        cp = zeit.content.cp.interfaces.ICenterPage(self)
-        for i, item in enumerate(content):
-            if predicate is None or predicate(item):
-                pop.append(item)
-                if self.hide_dupes and (
-                        cp.is_teaser_present_above(self.context, item)
-                        or cp.is_teaser_manual_below(self.context, item)):
-                    continue
-                result = item
-                break
-        for item in pop:
-            content.remove(item)
-        if result is None and not content:
-            # We've exhausted all available content due to duplicates, so we
-            # need to retrieve some more.
-            more_content = self._retrieve_content()
-            if more_content:
-                content[:] = more_content
-                return self._extract_newest(content, predicate)
-        return result
+    @property
+    def filter_query(self):
+        if not self.context.hide_dupes or not self.existing_teasers:
+            return lq.any_value()
+        return lq.not_(lq.or_(*[lq._field('uniqueId', '"%s"' % x.uniqueId)
+                                for x in self.existing_teasers]))
 
-    def select_modules(self, *interfaces):
-        for module in self.values():
-            if any([x.providedBy(module) for x in interfaces]):
-                yield module
 
+class ChannelContentQuery(SolrContentQuery):
 
-def is_lead_candidate(content):
-    metadata = zeit.cms.content.interfaces.ICommonMetadata(content, None)
-    if metadata is None:
-        return False
-    return metadata.lead_candidate
+    grok.name('channel')
+
+    SOLR_FIELD = {
+        'Channel': 'channels',
+        'Keyword': 'keywords',
+    }
+
+    def __init__(self, context):
+        super(SolrContentQuery, self).__init__(context)
+        self.query_string = self._build_query()
+        self.order = self.context.query_order
+
+    def _build_query(self):
+        query = zeit.find.search.query(filter_terms=[
+            zeit.solr.query.field_raw('published', 'published*')])
+        conditions = []
+        for type_, channel, subchannel in self.context.query:
+            if subchannel:
+                value = '%s*%s' % (channel, subchannel)
+            else:
+                # XXX Unclear whether this will work as desired for keywords.
+                value = '%s*' % channel
+            conditions.append(zeit.solr.query.field_raw(
+                self.SOLR_FIELD[type_], value))
+        if conditions:
+            query = zeit.solr.query.and_(
+                query, zeit.solr.query.or_(*conditions))
+        return query
+
+
+class CenterpageContentQuery(ContentQuery):
+
+    grok.name('centerpage')
+    # XXX If zeit.web wanted to implement pagination for CP queries, we'd have
+    # to walk over the *whole* referenced CP, which can be rather expensive.
```

Comment:
Ah verstehe. Der erste Teil (sehr aufwändig) steht ja auch im Kommentar. Das mit "brauchen wir bisher nicht und wird ignoriert" fehlt. Sollte im Kommentar ergänzt werden. Dann kann man für den Kommentar das explizite `total_hits = NotImplemented` auch stehen lassen. Es sollte aber dennoch zusätzlich auf die Basis-Klasse. (Ich finde es sauberer, wenn man schon in der Basis-Klasse sieht, was man überschreiben muss, damit man das Interface erfüllt.)